document kube-apiserver identity (#24921)
kubernetes/website: content/en/docs/tasks/administer-cluster/highly-available-master.md

To allow etcd clustering, ports needed to communicate between etcd instances will be opened (for inside cluster communication).
To make such deployment secure, communication between etcd instances is authorized using SSL.

### API server identity

{{< feature-state state="alpha"  for_k8s_version="v1.20" >}}

The API Server Identity feature is controlled by a feature gate
and is not enabled by default.  See
[Feature Gates](/docs/reference/command-line-tools-reference/feature-gates/)
for a general explanation of feature gates and how to enable and disable them.  The

> name of the feature gate is "APIServerIdentity".  You can enable API Server Identity

Reviewer: Hi there, please don't forget to add `APIServerIdentity` to https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/ and a short description of what functionality it provides. thank you!

Author: Done. Copied the one-sentence descriptions from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/features/kube_features.go

> The API Server Identity feature is controlled by a feature gate
> and is not enabled by default.  See
> [Feature Gates](/docs/reference/command-line-tools-reference/feature-gates/)
> for a general explanation of feature gates and how to enable and disable them.  The
> name of the feature gate is "APIServerIdentity".  You can enable API Server Identity
> by adding the following command-line flag to your `kube-apiserver` invocation:

Reviewer 2: I'd write this as:
```suggestion
The API Server Identity feature is controlled by a
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
and is not enabled by default. You can activate API Server Identity by enabling
the feature gate named `APIServerIdentity` when you start the
{{< glossary_tooltip text="API Server" term_id="kube-apiserver" >}}:
```

Reviewer 2: or:
```suggestion
_API Server identity_ is behind a
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
named `APIServerIdentity`, that is not enabled by default.
You can enable API Server Identity by adding the following command-line
flag to your `kube-apiserver` invocation:
```

Author: Thanks. Updated the paragraph with the first suggestion.

```shell
kube-apiserver \
--feature-gates=APIServerIdentity=true \
 # …and other flags as usual
```

> Each API server will be assigned with an unique ID on bootstrap. The list of

Reviewer 2: ```suggestion
During bootstrap, NAME_OF_THING assigns a unique ID to each API server. Each kube-apiserver
manages a [Lease](/docs/reference/generated/kubernetes-api/{{< param "version" >}}//#lease-v1-coordination-k8s-io) 
in the _kube-apiserver-lease_ {{< glossary_tooltip text="namespaces" term_id="namespace">}}.
The Lease contains the unique ID for the kube-apiserver. If another kube-apiserver tries to join
same control plane, the additional API server [picks a unique ID that is not used within the cluster?]

Enabling this feature [provides benefits?].

Each API server also exposes its own ID as a label dimension in its metrics: `TBD`.
```

Author: Updated the mechanism and the benefit. 

The metrics didn't made it into alpha. Dropped the sentence.

To make such deployment secure, communication between etcd instances is authorized using SSL.

### API server identity

{{< feature-state state="alpha"  for_k8s_version="v1.20" >}}

The API Server Identity feature is controlled by a
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
and is not enabled by default. You can activate API Server Identity by enabling
the feature gate named `APIServerIdentity` when you start the
{{< glossary_tooltip text="API Server" term_id="kube-apiserver" >}}:

```shell
kube-apiserver \
--feature-gates=APIServerIdentity=true \
 # …and other flags as usual
```

During bootstrap, each kube-apiserver assigns a unique ID to itself. The ID is
in the format of `kube-apiserver-{UUID}`. Each kube-apiserver creates a
[Lease](/docs/reference/generated/kubernetes-api/{{< param "version" >}}//#lease-v1-coordination-k8s-io)
in the _kube-system_ {{< glossary_tooltip text="namespaces" term_id="namespace">}}.
The Lease name is the unique ID for the kube-apiserver. The Lease contains a
label `k8s.io/component=kube-apiserver`. Each kube-apiserver manages deleting

> identity leases for kube-apiservers that are gone.

Reviewer 3: Shall we also introduce the lease duration and gc period flags?

Author: added the flags

> identity leases for kube-apiservers that are gone.

Author: updated with details. Thanks

> Enabling this feature is a prerequisite for using features that involve HA API
> server coodination (e.g. the StorageVersion API feature).

Reviewer 2: nits (we avoid Latin abbreviations; typo fix)
```suggestion
Enabling this feature is a prerequisite for using features that involve HA API
server coordination (for example, the StorageVersion API feature).
```

Is StorageVersion a `kind` or the name of a feature gate? I'm not sure readers can work that out.

Author: it's the feature gate. Updated
